diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -349,9 +349,58 @@ int DifferentialFunctionComparator::cmpAllocs(const CallInst *CL,
            || STyL->getStructName() != STyR->getStructName();
 }
 
-/// Check if the given operation can be ignored (it does not affect semantics).
-bool mayIgnore(const Instruction *Inst) {
-    return isa<AllocaInst>(Inst) || Inst->isCast();
+/// Check if the given operation can be ignored (it does not affect semantics)
+/// for control flow only diffs.
+bool DifferentialFunctionComparator::mayIgnore(const Instruction *Inst) const {
+    if (controlFlowOnly)
+        return isa<AllocaInst>(Inst) || Inst->isCast();
+    else {
+        if (isa<AllocaInst>(Inst))
+            return true;
+        if (auto Cast = dyn_cast<CastInst>(Inst)) {
+            if (auto StrTy = dyn_cast<StructType>(Cast->getSrcTy())) {
+                if (StrTy->hasName() && StrTy->getName().startswith("union"))
+                    // Casting from a union type is safe to ignore, because in
+                    // case it was generated by something different than a
+                    // replacement of a simple type with a union, the
+                    // comparator would detect a difference when the value is
+                    // used, since it would have a different type.
+                    return true;
+            }
+
+            if (Cast->getSrcTy()->isPointerTy()
+                && Cast->getDestTy()->isPointerTy())
+                // Pointer cast is also safe to ignore - in case there will be
+                // an instruction affected by this (i.e. a store, load or GEP),
+                // there would be an additional difference that will be detected
+                // later.
+                return true;
+
+            if (Cast->getSrcTy()->isIntegerTy()
+                && Cast->getDestTy()->isIntegerTy()) {
+                // Integer-to-integer casts are safe to ignore only if two
+                // conditions are met:
+                // 1. The destination type is larger than the source type.
+                // 2. No arithmetic is performed on the number.
+                // Note: store instructions aren't a problem here, since the
+                // types of the arithmetic instructions would be different if
+                // there is arithmetic performed on the value after being
+                // loaded back.
+                auto IntTyL = dyn_cast<IntegerType>(Cast->getSrcTy());
+                auto IntTyR = dyn_cast<IntegerType>(Cast->getDestTy());
```

Comment:
I'd rather call these variables `IntTySrc` and `IntTyDest` because we use `L` and `R` suffices when comparing things between modules, which is not the case here.